src/azul/collections.py

```diff
@@ -271,3 +273,94 @@ def to_dict(self):
             k: v.to_dict() if isinstance(v, NestedDict) else v
             for k, v in self.items()
         }
+
+
+class OrderedSet(MutableSet[K]):
+    """
+    A mutable set that maintains insertion order. Unlike similar implementations
+    of the same name that are floating around on the internet, it is not a
+    sequence.
+
+    >>> s = OrderedSet(['b', 'a', 'c', 'b'])
+    >>> s
+    OrderedSet(['b', 'a', 'c'])
+
+    >>> s.discard('a')
+    >>> s
+    OrderedSet(['b', 'c'])
+
+    >>> s.add('a')
+    >>> s
+    OrderedSet(['b', 'c', 'a'])
+    """
+
+    def __init__(self, members: Iterable[K] = (), /) -> None:
+        self.inner: dict[K, None] = dict.fromkeys(members)
+
+    def __repr__(self) -> str:
+        contents = repr(list(self)) if self else ''
+        return f'{type(self).__name__}({contents})'
+
+    def __iter__(self) -> Iterator[K]:
+        return iter(self.inner)
+
+    def __len__(self) -> int:
+        return len(self.inner)
+
+    def __eq__(self, other: Any) -> bool:
+        """
+        >>> OrderedSet() == set()
+        True
+
+        >>> s1 = OrderedSet(['3', '1', '3'])
+        >>> s2 = {'1', '3', '1'}
+        >>> (s1 and s2) == (s2 and s1)
+        True
+
+        >>> (s2 or s1) == (s1 or s2)
+        True
+
+        >>> (s1 == s2) == (s2 == s1)
+        True
+
+        >>> s1 == s2 and s2 == s1
```

Comment:
This proves commutativity but when it fails, the failure message doesn't reveal which ordering of the operands cause the failure. Split into two assertions.